### PR TITLE
Fix cyclic reference errors in collection strawman

### PR DESCRIPTION
**Pull request by odersky**
This should make collection strawman #137 build. There were some "interesting" F-bounded types which triggered a series of problems when read back with Tasty unpicklers. First, the underlying error message ("wrong number of arguments") was hidden because it caused a CRE. Next, the wrong number of argument problems demanded that we are more careful reading applicatons. Once that was done we got Stackoverflows. To avoid these, we have to maintain any LazyRefs we have in the original types in the pickle information. 

Review by @szeiger or @julienrf?

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala [hunk 1]
```diff
@@ -293,6 +293,11 @@ class TreeUnpickler(reader: TastyReader, nameAtRef: NameRef => TermName, posUnpi
           RecType(rt => registeringType(rt, readType()))
         case RECthis =>
           RecThis(readTypeRef().asInstanceOf[RecType])
+        case LAZYref =>
+          val rdr = fork
+          def readUnderlying() = rdr.readType()
+          skipTree()
+          LazyRef(readUnderlying)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala]: This means LazyRefs will keep the context from which they were unpickled around, if these LazyRefs are subsequently stored in some cache, that could lead to space leaks, no?

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala [hunk 2]
```diff
@@ -708,10 +709,15 @@ object messages {
...
-      case symbol   => symbol.showFullName
-    }
+    private val prettyName =
+      try
+        fntpe.termSymbol match {
+          case NoSymbol => fntpe.show
+          case symbol   => symbol.showFullName
+        }
+      catch {
+        case NonFatal(ex) => fntpe.show
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: If this is only for cyclic references, could we do `case ex: CyclicReference` instead?

File: compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala (hunk 1 above)
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala]: Yes, correct. Maybe the best option is to allow a context to be passed into a LazyRef. In the case of unpickling, this is fine; we don't need to hang on to more than the reader (I think, at least that's what we do for the other completers).

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala (hunk 2 above)
Comment (odersky) [inline compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: I think we by default we should not throw any exception from printing. It detracts from the underlying cause.

Comment (smarter) [inline compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: Sure, but we shouldn't be throwing any kind of exceptions, and silently catching them all might hide issues.

**Comment by DarkDimius** [comment]
@odersky, instead of pickling lazy refs, could we create them at un-pickling similarly to how we create them at typer?

**Comment by odersky** [comment]
> @odersky, instead of pickling lazy refs, could we create them at un-pickling similarly to how we create them at typer?

I tried that, but it didn't help. Did not get to find out why.

**Comment by odersky** [comment]
Merging now to get the strawman unstuck, but I opened an issue to revert the LazyRef addition if possible. /cc @julienrf

**Comment by julienrf** [comment]
Thanks @odersky!

**Pull request merged**